### DISCUSSION TITLE
store family cache on provider

### DISCUSSION THREAD

**Discussion by neiker** [pmndrs/jotai#1852] [Ideas] [answered]
Hey everyone! I searched on history but don't seems to be a post on this topic. Please let me know if it's duplicated.

I found it confusing that the cache for `atomFamily` it's stored globally but the value for `atom` it's stored on the closest provider.  It might create situations when you think unmounting the provider will trigger GC for all your state and mounting a new provider will create an entirely new state but in reality family's cache it's the same.

This is especially problematic for react native. Imagine you have a chat app, with a `/chat/:id` route and inside that screen, a `<JotaiProvider>`. You have different atoms for the state and an `atomFamily` for messages (potentially with thousands of messages). Each time you navigate to the screen the messages are added to the family, when you navigate back to the previous screen and open another chat more messages are added and so on. 

I explored 2 solutions for this:

 - Call `family.setShouldRemove(...` when screen it's unmounted
 - Wrap atomFamily in an atom
```ts
const fooAtomFamilyAtom = atom(() =>
  atomFamily(
    ...
  )
);
```

Both with disadvantages, first one requires you to remember to perform the clean up while on second one cleanup it's managed for you but it adds an extra step to use the atomFamily:
```ts
const atomFamily = useAtomValue(fooAtomFamilyAtom);
const [item, setItem] = useAtom(atomFamily({ id }));
```

and:
```ts
get(get(fooAtomFamilyAtom)({ id }))
```

Now I'm solving it by locally adding `atomFamilyAtom()` and `useAtomFamilyAtom(atomFamilyAtom, param)` but seems like something it should be either the default behaviour for `atomFamily`, an opt-in or at least mentioned on the docs

**Comment by dai-shi** [comment]
> I found it confusing that the cache for `atomFamily` it's stored globally but the value for `atom` it's stored on the closest provider. It might create situations when you think unmounting the provider will trigger GC for all your state and mounting a new provider will create an entirely new state but in reality family's cache it's the same.

atomFamily doesn't hold atom values. So, your assumption should be correct. If you unmount the provider, atom values will be GC'd. atomFamily only caches atom config objects. I suspect you are using the default store (no Provider) somewhere.

**Reply by neiker** [reply]
I see, thanks! I might be misinterpreting something. 

In my case the memoize it's returning the same atom config even when the providers it's different. [Here's](https://codesandbox.io/s/nervous-albattani-82q39n?file=/src/App.tsx:2241-2515) my issue replicated on one of the examples on the showcase. With unique ids is hard to spot so I changed id to incremental ids

https://user-images.githubusercontent.com/688444/227880126-2cfdcd28-0c72-466a-b4f5-f962614d4813.mov

**Reply by dai-shi** [reply, marked as answer]
> it's returning the same atom config even when the providers it's different.

Yes, that's the goal of `atomFamily`.
Having the same atom config, doesn't mean to have the same value for different providers.

The issue in your case is that you use the same id, which returns the same config which is initialized with different param.
It's not the issue of `atomFamily`. We could reproduce such a behavior without `atomFamily`.

That said, I think you want to try "atoms-in-atom" pattern, which would work more intuitively for you.
